extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java

```diff
@@ -119,6 +125,20 @@ public WindowOperatorQueryFrameProcessor(
     for (int i = 0; i < frameReader.signature().size(); i++) {
       typeStrategies[i] = frameReader.signature().getColumnType(i).get().getNullableStrategy();
     }
+
+    // Get virtual columns to be added to the frame writer.
+    this.partitionBoostVirtualColumn = new SettableLongVirtualColumn(QueryKitUtils.PARTITION_BOOST_COLUMN);
+    final List<VirtualColumn> frameWriterVirtualColumns = new ArrayList<>();
+    frameWriterVirtualColumns.add(partitionBoostVirtualColumn);
+
+    final VirtualColumn segmentGranularityVirtualColumn =
+        QueryKitUtils.makeSegmentGranularityVirtualColumn(jsonMapper, query);
+
+    if (segmentGranularityVirtualColumn != null) {
+      frameWriterVirtualColumns.add(segmentGranularityVirtualColumn);
+    }
+
+    this.frameWriterVirtualColumns = VirtualColumns.create(frameWriterVirtualColumns);
```

Comment:
I think this should be remove from here as it seems like a cross cutting aspect
The same lines are present in other `QueryKit` classes

could you refactor these changes out from all `QueryKit` classes and place them at a common place?

Reply:
These lines are same in `ScanQueryFrameProcessor`, but they are different in `GroupByPostShuffleFrameProcessor`, and not present in `GroupByPreShuffleFrameProcessor`. So I don't see how to cleanly create a base class for use by all frame processors.

Also, it feels unrelated to this PR changes, so we should tackle it separately in my opinion. Would like to keep the changes affect only the window-related layers if possible.